### ISSUE TITLE
Replacement of `max` like functions in the model

### ISSUE THREAD

**Issue by otthii** [dragonfly-science/seabird-risk-assessment#23]
In the current master at the point of [`d45a88f`](https://github.com/dragonfly-science/seabird-risk-assessment/commit/d45a88f5ca7aca7305d3ca80e5b850263bb92b96), `model/jags/model.bug` and `model/stan/model.stan`, `max` (or `fmax`) functions to choose nearly zero value or product of parameters are used for expected value of Poisson distribution like below;
> DEAD_CAPTURES_O[i] ~ poisson(fmax(1E-8, mu_observable_dead_captures_o[i]));

I'd like to explain the problem to use `max` function and suggest a solution to avoid the problem.
I'm glad to receive what do you think about this suggestion.

### Problem
- It is a good technique to avoid having negative values on the expected value on Poisson distribution which doesn't accept negative values, however, using `max` like functions into model structure will causes heavily biased posteriors on the MCMC algorithm. 
- In the process of MCMC sampling, candidates of posterior are suggested based on gradient of likelihood function which are made of data and model structure. If we force to change suggested parameters in the process, it becomes not a manner of MCMC algorithm and Bayes Inference. 

### Suggestion to solve the problem
It is highly required to remove such function and I recommend to use "link functions" instead of that.
- Concept of link function is generally accepted in the procedure of generalized liner model (GLM; `stat::glm` on R).
- To express the relationships of explanatory variables as a liner equation, expected values are transformed suitable forms such as `mu' = beta * X + alpha` (`mu'` is a transformed expected value, X is explanatory variable and the others are parameters) 
- For Poisson distribution, log link function is usually applied and it seems also suitable in this case.
- log link function transforms `mu` to `log(mu)` then each explanatory variables are connected by liner-style such as `log(mu) =  beta * X + alpha` and `Y ~ poisson(mu)`.
- With using log-link functions, we don't have to concern about negative values and variable range of the parameters.

When we look at our model, the part of Poisson distribution in `model.stan` is now described like below;
```
q_o = q0 * (q_g_o .* q_f_o .* q_gf_o);
mu_observable_incidents_o = (q_o .* OVERLAP_O .* ntot[SPECIES_O]) * p_observable
mu_observable_live_incidents_o = (mu_observable_incidents_o .* p_live_cap[FISHING_GROUP_O]) * p_identified;
LIVE_CAPTURES_O[i] ~ poisson(fmax(1E-8, mu_observable_live_incidents_o[i]));
```

Now the part can be changed according to the concept of log-link function as below;

```
q_o = q0 + (q_g_o + q_f_o + q_gf_o);
mu_observable_incidents_o = (q_o + OVERLAP_O + ntot[SPECIES_O]) + p_observable;
mu_observable_live_incidents_o = mu_observable_incidents_o + p_live_cap[FISHING_GROUP_O]) + p_identified;
LIVE_CAPTURES_O[i] ~ poisson_log(mu_observable_live_incidents_o[i]);
```
In Stan, some of link function is automatically transformed with using specific function like `poisson_log`. 

It is very useful but we have to taking care about outputted parameters because each parameters are also transformed so it need back-calculation to get actual values.

Thank you very much for reading to the end.

**Comment by edwardabraham** [comment]
I would be happy for it to be re-written in this way. However, I personally find it more straightforward to work in natural units (rather than log transformed).  The transforming makes it easy to introduce bugs (by forgetting a transform somewhere). That said, I certainly take the point about `fmax` not being ideal. There are two sources of problems: 
1. The prior for the probability parameters (such as `p_identified`) includes zero, and very small values may be rounded to zero
2. At the moment, the overlap can be zero for a particular fishery and species group

The first problem can be solved by setting a lower bound in the definition of the  parameters (such as
`real<lower=1E-12> q0;`), so that the region which rounds to zero is excluded. I am comfortable with this, as I do not expect any of these parameters to have values so close to zero. I like the probabilities having uniform `beta(1, 1)` priors, although a `beta(2, 2)` distribution excludes 0 and 1 and is peaked around 0.5, and could be a possibility. The `q`'s have log-normal priors, which are appropriate and which exclude zero.

The second problem can be solved by removing the rows with zero overlap during preparation of the model data (assuming that no birds are observed caught where there is zero overlap!).  A second solution would be to  still use `fmax` where the overlap appears, e.g.,
 `fmax(1E-12, OVERLAP_O)`, so that there are no strict zeros. In practice this is interpreted as saying that even though there was no overlap between the birds and the fisheries, there may be a very small possibility of recording a bird as being caught. In practice we have found that this occurs because of errors in the seabird distributions or weird records put down by observers. This would be equivalent to having a small probability of finding a bird anywhere. It is possible that if (2) is solved, then (1) won't be a problem.

I think that with a little care, the need for `fmax` can be removed, without needing much re-writing.

If you would still like to re-write it on the log scale, however, then i would prefer that we keep the same meaning for the parameters, so that the log appears on the right hand side as well, e.g.:
```
transformed parameters {
...
log_q_o = log(q0)  +  log(q_g_o ) + log(q_f_o)   + log(q_gf_o);
log_mu_observable_incidents_o = log(q_o)  +  log(fmax(1E-12, OVERLAP_O))   + log(ntot[SPECIES_O])) + log(p_observable);
log_mu_observable_live_incidents_o = log_mu_observable_incidents_o 
 + log(p_live_cap[FISHING_GROUP_O])) + log(p_identified)
...
}

model {
...
LIVE_CAPTURES_O[i] ~ log_poisson(log_mu_observable_live_incidents_o[i]));
...
}
```

**Comment by edwardabraham** [comment]
I am going to have a go over the next 10-minutes at removing the need for `fmax` anywhere except around `OVERLAP_O`.

**Comment by edwardabraham** [comment]
I have pushed a commit  608b385 that removes `fmax` everywhere, except for keeping the observed overlap above 1E-12 in every row. I recommend that these rows are removed in preparation of the model data and I think that @yvanrichard will look at this now.

**Comment by otthii** [comment]
I completely agree with you about concerns 1 and 2 because I've run my re-written model yesterday on my commit [`58cf1e7`] (https://github.com/dragonfly-science/seabird-risk-assessment/commit/58cf1e760520c125c5bfbc79a0006ee7ec3899de) and encountered similar problem. 

About probability parameters, I believe it can be solved with separating to another process. I'd like to discuss about the matter in another issue.

I feel that the problem of zero values of 'OVERLAP_O' is more complex. When I looked at test data, there were very high correlation between zero catch and zero overlap. This problem is very annoying. I have several ideas to clear this problem but it is not easy anyway. Please give me another time to thinking....

I think your suggested idea is nice to solve the problem at this time. Thank you very much for your great work!
